tools/generator/templates/player/caml/interface.cc.jinja2

```diff
@@ -0,0 +1,236 @@
+{
+    CAMLparam1(in);
+    CAMLreturnT(std::string, String_val(in));
+}
+template <>
+int caml_to_cxx<value, int>(value in)
+{
+    CAMLparam1(in);
+    CAMLreturnT(int, Int_val(in));
+}
+template <>
+double caml_to_cxx<value, double>(value in)
+{
+    CAMLparam1(in);
+    CAMLreturnT(double, Double_val(in));
+}
+template <>
+bool caml_to_cxx<value, bool>(value in)
+{
+    CAMLparam1(in);
+    CAMLreturnT(bool, Int_val(in));
+}
+template <typename Cxx>
+std::vector<Cxx> caml_to_cxx_array(value in)
+{
+    CAMLparam1(in);
+    std::vector<Cxx> out;
+    mlsize_t size = Wosize_val(in);
+    for (size_t i = 0; i < size; ++i)
+        out.push_back(caml_to_cxx<value, Cxx>(Field(in, i)));
+    CAMLreturnT(std::vector<Cxx>, out);
+}
+{% for enum in game.enum %}
+
+{{ enum.enum_summary|cxx_comment }}
+template<>
+value cxx_to_caml<value, {{ enum.enum_name }}>({{ enum.enum_name }} in)
+{
+    CAMLparam0();
+    CAMLreturn(Val_int(in));
+}
+
+template<>
+{{ enum.enum_name }} caml_to_cxx<value, {{ enum.enum_name }}>(value in)
+{
+    CAMLparam1(in);
+    CAMLreturnT({{ enum.enum_name }}, ({{ enum.enum_name }})Int_val(in));
+}
+{% endfor %}
+{% for struct in game.struct %}
+
+{{ struct.str_summary|cxx_comment }}
+template<>
+value cxx_to_caml<value, {{ struct.str_name }}>({{ struct.str_name }} in)
+{
+    CAMLparam0();
+    CAMLlocal1(out);
+    out = caml_alloc({{ struct.str_field|length }}, 0);
+    {% for field_name, field_type, field_comment in struct.str_field %}
+    caml_initialize(&Field(out, {{ loop.index0 }}), {{ field_type|cxx_to_caml}}(in.{{ field_name }}));
+    {% endfor %}
+    CAMLreturn(out);
+}
+
+template<>
+{{ struct.str_name }} caml_to_cxx<value, {{ struct.str_name }}>(value in)
+{
+    CAMLparam1(in);
+    {{ struct.str_name }} out;
+    {% for field_name, field_type, field_comment in struct.str_field %}
+    out.{{ field_name }} = {{ field_type|caml_to_cxx }}(Field(in, {{ loop.index0 }}));
+    {% endfor %}
+    CAMLreturnT({{ struct.str_name }}, out);
+}
+{% endfor %}
+
+/*
+** Inititialize caml
+*/
+static inline void _init_caml()
+{
+    static bool is_initialized = false;
+    if (!is_initialized)
+    {
+        is_initialized = true;
+        const char* argv[2] = {"./caml", NULL};
+        caml_startup(const_cast<char**>(argv));
+    }
+}
+{% for func in game.function %}
+{%- set arg_list = func.fct_arg|cxx_nonempty_or_unit %}
+
+{{ func.fct_summary|cxx_comment }}
+extern "C" value ml_{{ func.fct_name }}({# argument list #}
+                                        {%- set comma = joiner(', ') -%}
+                                        {%- for arg_name, _, _ in arg_list -%}
+                                        {{ comma() }}value {{ arg_name }}
+                                        {%- endfor -%}
+                                       )
+{
+    CAMLparam0();
+    {# TODO: forbid instanciation if there's more than 5 parameters, OCaml
+     # has CAMLxparam0 to CAMLxparam5 only #}
```

Comment:
TODO should be "Use CAMLparamN, then CAMLxparamN, by grouping the arguments 5 by 5"